### PR TITLE
Extract the task/engine contract from DualPipeV

### PR DESCRIPTION
**Pull request by haok1402**
## Extract the task/engine contract from DualPipeV

`step()` takes a list of `Microbatch` plus a task-owned `objective`, and returns that objective's per-micro-batch output instead of a loss. Micro-batches may differ in shape, which RL rollouts need. Every pipeline rank gets the same micro-batches and derives its own activation shapes, so no per-step metadata broadcast or host sync remains. `comm.py` is merged into `dualpipev.py`.

### Why

`step()` took a `criterion` and `labels`, so supervised pretraining was the only job it could express. RL needs a forward-only pass with no loss, objectives over arbitrary per-micro-batch data, and micro-batches of differing length — none expressible against a hardcoded criterion and a process-global shape constant.

### Correctness

Base runs twice per mesh. The two base runs are the anchors; feature is measured against both. Distances are mean |Δloss| over 32 steps.

| mesh | base | base2 | feat | base–base2 | base–feat | base2–feat |
|---|---|---|---|---|---|---|
| qwen3-30b-a3b pp2 ep8 | 9.5223 | 9.5094 | 9.5356 | 0.0128 | 0.0373 | 0.0480 |
| qwen3-30b-a3b pp4 ep4 | 9.4870 | 9.5030 | 9.5004 | 0.0086 | 0.0097 | 0.0110 |
| deepseek-v2-lite pp2 ep4 dp2 | 8.4714 | 8.4423 | 8.4581 | 0.0312 | 0.0078 | 0.0249 |
| deepseek-v2-lite pp2 ep4 cp2 | 8.2015 | 8.1984 | 8.1936 | 0.0163 | 0.0096 | 0.0100 |

- Step 1 bit-identical across all legs on every mesh.
- feat's final loss lies between the two base runs on 2 of 4 meshes.
- feat-to-anchor distance is below anchor-to-anchor distance on 2 of 4.

12 runs × 32 steps, fresh weights, linear warmup 1e-6 → 1e-5, seq 4096, global batch 512. pp>1 everywhere (the shape logic is inert at pp=1); pp4 for intermediate stages, dp2 for the data path now running on every rank, cp2 for the zigzag split. All legs of a mesh shared nodes except deepseek pp2 ep4 cp2.

| mesh | base | base2 | feat |
|---|---|---|---|
| qwen3-30b-a3b pp2 ep8 | [2tu78ld3](https://wandb.ai/PithTrain/pr-87/runs/2tu78ld3) | [g03amk91](https://wandb.ai/PithTrain/pr-87/runs/g03amk91) | [ck2nhrqz](https://wandb.ai/PithTrain/pr-87/runs/ck2nhrqz) |
| qwen3-30b-a3b pp4 ep4 | [8urled47](https://wandb.ai/PithTrain/pr-87/runs/8urled47) | [5b2ivsbn](https://wandb.ai/PithTrain/pr-87/runs/5b2ivsbn) | [i3clc53f](https://wandb.ai/PithTrain/pr-87/runs/i3clc53f) |
| deepseek-v2-lite pp2 ep4 dp2 | [vryxpdyc](https://wandb.ai/PithTrain/pr-87/runs/vryxpdyc) | [uehdnedp](https://wandb.ai/PithTrain/pr-87/runs/uehdnedp) | [t51c92wj](https://wandb.ai/PithTrain/pr-87/runs/t51c92wj) |
| deepseek-v2-lite pp2 ep4 cp2 | [rwb2smm7](https://wandb.ai/PithTrain/pr-87/runs/rwb2smm7) | [u7tltuy2](https://wandb.ai/PithTrain/pr-87/runs/u7tltuy2) | [cbl6c6hk](https://wandb.ai/PithTrain/pr-87/runs/cbl6c6hk) |

All runs: https://wandb.ai/PithTrain/pr-87

### Performance

Peak memory, base / base2 / feat: 73.72 / 73.67 / 73.40 GB on qwen3 pp2 ep8; 50.24 / 50.10 / 50.19 GB on deepseek pp2 ep4 dp2. Step-time medians agree within 1% and are unordered between legs; these use real routing, so a force-balanced (`benchmark=True`) pass is still pending.

Metadata path at 1024 micro-batches: 42.5µs before the change, 185.3µs with the shape broadcast, 39.9µs once every rank derives shapes locally.

### Tests

`tests/test_dualpipev.sh` passes 8/8 on deepseek-v2-lite, qwen3-30b-a3b and gpt-oss-20b across dense/packed × uniform/ragged, two `step()` calls each so shape changes between steps are covered.

### CODE REVIEW

**Comment by haok1402** [comment]
@claude review

File: pithtrain/dualpipe/dualpipev.py [hunk 1]
```diff
@@ -404,15 +428,38 @@ def _free_tensors(self) -> None:
...
+                dtype=training.PARAM_DTYPE,
+                device=distributed.device,
+                requires_grad=True,
+            )
+            for shape in self.p2p_shapes[chunk_id]
+        ]
+        src = dist.distributed_c10d.get_global_rank(self.pp_group, src)
+        for tensor in tensors:
+            if tensor is not None:
+                self.comm_ops.append(dist.P2POp(dist.irecv, tensor, src))
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/dualpipe/dualpipev.py]: @claude Great catch. I've applied the fix locally and pushed.

File: pithtrain/tasks/pretrain_lm.py [hunk 2]
```diff
@@ -109,17 +112,37 @@ def get_global_batch(
     local_tokens = local_tokens.to(device, non_blocking=True)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/tasks/pretrain_lm.py:112]: @claude Correct, and it is a deliberate trade: handing every rank the same micro-batches is what lets each derive its own activation shapes locally, which removed three per-step collectives and both host syncs, and it matches how slime already hands every actor the same rollout data. The cost is ~1-2 MB per rank, which is trivial.
